[hardhat-chai-matchers] Edge case with `not.to.be.reverted` and view functions

### Version of Hardhat

hardhat-chai-matchers 2.0.3 / hardhat-ethers 3.0.5

### What happened?

The (not) `reverted` matcher throws an error when calling a `view` function returning ``bytes32` that does not revert.

This is because the code inside `expect` actually returns a string, and `reverted.ts` interprets that return value as a hash. Then, when it tries to get the transaction receipt it fails at [this line](https://github.com/NomicFoundation/hardhat/blob/%40nomicfoundation/hardhat-chai-matchers%402.0.3/packages/hardhat-chai-matchers/src/internal/reverted/reverted.ts#L37).

Tested with latest tag `2.0.3`. Probably also happens with other types that can be interpreted as a hash?

### Minimal reproduction steps

Suppose I have library with a view function that doesn't revert, but returns `bytes32`:

Solidity:
```
contract Test {
    bytes32 private _value;

    function setValue(bytes32 value) external {
        _value = value;
    }

    function getValueBytes32() external view returns (bytes32) {
        return _value;
    }

    function getValueUint() external view returns (uint256) {
        return uint256(_value);
    }
}
```

Test:

```
it('is ok', async () => {
    await expect(test.getValueUint()).not.to.be.reverted;
});

it('throws', async () => {
    await expect(test.getValueBytes32()).not.to.be.reverted;
});
```

The error in the second scenario is:
```
HardhatChaiMatchersAssertionError: Assertion error: receipt should not be null
```

A simple workaround is to just not use `expect` in these cases:

```
it('is ok as a workaround', async () => {
    test.getValueBytes32(); // does not revert
});
```

But ideally `not.reverted` should work as with non-view functions and the other return values.

### Search terms

chai-matchers, not.to.be.reverted

--- a/.changeset/ninety-trainers-rush.md
+++ b/.changeset/ninety-trainers-rush.md
@@ -2,4 +2,4 @@
 "@nomicfoundation/hardhat-chai-matchers": patch
 ---
 
-Enhanced byte32 string in `reverted` matcher check
+Enhanced the `reverted` matcher to correctly handle `bytes32` strings (thanks @iosh!)

--- a/.changeset/ninety-trainers-rush.md
+++ b/.changeset/ninety-trainers-rush.md
@@ -1,0 +1,5 @@
+---
+"@nomicfoundation/hardhat-chai-matchers": patch
+---
+
+Enhanced byte32 string in `reverted` matcher check

--- a/packages/hardhat-chai-matchers/src/internal/reverted/reverted.ts
+++ b/packages/hardhat-chai-matchers/src/internal/reverted/reverted.ts
@@ -3,7 +3,11 @@ import type EthersT from "ethers";
 import { buildAssert } from "../../utils";
 import { REVERTED_MATCHER } from "../constants";
 import { assertIsNotNull, preventAsyncMatcherChaining } from "../utils";
-import { decodeReturnData, getReturnDataFromError } from "./utils";
+import {
+  decodeReturnData,
+  getReturnDataFromError,
+  parseBytes32String,
+} from "./utils";
 
 export function supportReverted(
   Assertion: Chai.AssertionStatic,
@@ -35,6 +39,14 @@ export function supportReverted(
         }
 
         const receipt = await getTransactionReceipt(hash);
+
+        if (receipt === null) {
+          // If the receipt is null, maybe the string is a bytes32 string
+          if (isByte32String(hash)) {
+            assert(false, "Expected transaction to be reverted");
+            return;
+          }
+        }
 
         assertIsNotNull(receipt, "receipt");
         assert(
@@ -133,4 +145,13 @@ function isTransactionReceipt(x: unknown): x is { status: number } {
 
 function isValidTransactionHash(x: string): boolean {
   return /0x[0-9a-fA-F]{64}/.test(x);
+}
+
+function isByte32String(v: string): boolean {
+  try {
+    parseBytes32String(v);
+    return true;
+  } catch {
+    return false;
+  }
 }

--- a/packages/hardhat-chai-matchers/src/internal/reverted/reverted.ts
+++ b/packages/hardhat-chai-matchers/src/internal/reverted/reverted.ts
@@ -42,7 +42,7 @@ export function supportReverted(
 
         if (receipt === null) {
           // If the receipt is null, maybe the string is a bytes32 string
-          if (isByte32String(hash)) {
+          if (isBytes32String(hash)) {
             assert(false, "Expected transaction to be reverted");
             return;
           }
@@ -147,7 +147,7 @@ function isValidTransactionHash(x: string): boolean {
   return /0x[0-9a-fA-F]{64}/.test(x);
 }
 
-function isByte32String(v: string): boolean {
+function isBytes32String(v: string): boolean {
   try {
     parseBytes32String(v);
     return true;

--- a/packages/hardhat-chai-matchers/src/internal/reverted/utils.ts
+++ b/packages/hardhat-chai-matchers/src/internal/reverted/utils.ts
@@ -125,3 +125,8 @@ export function resultToArray(result: EthersT.Result): any[] {
         : x
     );
 }
+
+export function parseBytes32String(v: string): string {
+  const ethers = require("ethers") as typeof EthersT;
+  return ethers.decodeBytes32String(v);
+}

--- a/packages/hardhat-chai-matchers/test/reverted/reverted.ts
+++ b/packages/hardhat-chai-matchers/test/reverted/reverted.ts
@@ -112,6 +112,14 @@ describe("INTEGRATION: Reverted", function () {
           "Expected a valid transaction hash, but got '0x123'"
         );
       });
+
+      it("promise of an byte32 string", async function () {
+        await expect(
+          Promise.resolve(
+            "0x3230323400000000000000000000000000000000000000000000000000000000"
+          )
+        ).not.to.be.reverted;
+      });
     });
 
     describe("with a TxResponse as its subject", function () {

